posix/fdtable.c

```diff
@@ -0,0 +1,341 @@
+{
+    oe_atexit(_free_table);
+}
+
+static int _resize_table(size_t new_size)
+{
+    int ret = -1;
+
+    /* The fdtable cannot be bigger than the maximum int file descriptor. */
+    if (new_size > OE_INT_MAX)
+        goto done;
+
+    if (oe_once(&_once, _install_atexit_handler_once) != OE_OK)
+        goto done;
+
+    /* Round the new capacity up to the next multiple of the chunk size. */
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > OE_INT_MAX)
+        new_size = OE_INT_MAX;
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused porition. */
+        memset(p + _table_size, 0, (n - _table_size) * sizeof(entry_t));
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static void _list_remove(list_t* list, entry_t* entry)
+{
+    if (entry->prev)
+        entry->prev->next = entry->next;
+    else
+        list->head = entry->next;
+
+    if (entry->next)
+        entry->next->prev = entry->prev;
+    else
+        list->tail = entry->prev;
+
+    entry->prev = NULL;
+    entry->next = NULL;
+    entry->is_on_list = true;
+}
+
+static void _list_prepend(list_t* list, entry_t* entry)
+{
+    if (list->head)
+    {
+        entry->prev = NULL;
+        entry->next = list->head;
+        list->head->prev = entry;
+        list->head = entry;
+    }
+    else
+    {
+        entry->next = NULL;
+        entry->prev = NULL;
+        list->head = entry;
+        list->tail = entry;
+    }
+
+    entry->is_on_list = true;
+}
+
+/*
+**==============================================================================
+**
+** Public interface:
+**
+**==============================================================================
+*/
+
+int oe_fdtable_assign(oe_device_t* device)
+{
+    int ret = -1;
+    bool locked = false;
+    entry_t* entry;
+
+    oe_spin_lock(&_lock);
+    locked = true;
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Reserve space for stdin, stdout, and stderr on the first call. */
+    if (_table_size < OE_STDERR_FILENO + 1)
+    {
+        if (_resize_table(OE_STDERR_FILENO + 1) != 0)
+        {
+            oe_errno = OE_ENOMEM;
+            goto done;
+        }
+    }
+
+    /* Get an entry either from the free list or from the table. */
+    if (_free_list.head)
+    {
+        entry = _free_list.head;
+        _list_remove(&_free_list, entry);
+    }
+    else
+    {
+        if (_next == _table_size && _resize_table(_next + 1) != 0)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+
+        entry = &_table[_next++];
+    }
+
+    /* Initialize the entry. */
+    entry->device = device;
+
+    /* Set the file descriptor return value. */
+    ret = (int)(entry - _table);
+
+done:
+
+    if (locked)
```

Comment:
Maybe you dont need the locked flag since all paths would ensure it would be locked.

Reply:
Thanks Radhika. Fixed.